task/backend/executor/executor.go

```diff
@@ -126,15 +126,18 @@ func (p *syncRunPromise) Cancel() {
 
 func (p *syncRunPromise) finish(res *runResult, err error) {
 	p.finishOnce.Do(func() {
-		defer p.logEnd()
+		defer func() {
+			// Always cancel p's context.
+			// If finish is called before p.qs.Query completes, the query will be interrupted.
+			// If afterwards, then p.cancel is just a resource cleanup.
+			p.cancel()
+
+			p.logEnd()
 
-		// Always cancel p's context.
-		// If finish is called before p.qs.Query completes, the query will be interrupted.
-		// If afterwards, then p.cancel is just a resource cleanup.
-		defer p.cancel()
+			close(p.ready)
+		}()
```

Comment:
I changed this because I found a race condition here with the `logEnd()` happening after `close(p.ready)`.
This mean the called to `promise.Wait()` would unblock before the final log was made.